packaging/tags.py

```diff
@@ -200,53 +245,98 @@ def _generic_abi():
-    # type: (PythonVersion, str, str, Iterable[str]) -> Iterator[Tag]
-    for tag in (Tag(interpreter, abi, platform) for platform in platforms):
-        yield tag
-    for tag in (Tag(interpreter, "none", platform) for platform in platforms):
+def pypy_tags(
+    interpreter=None,  # type: Optional[str]
+    abis=None,  # type: Optional[Iterable[str]]
+    platforms=None,  # type: Optional[Iterable[str]]
+):
+    # type: (...) -> Iterator[Tag]
+    """
+    Yields the tags for a PyPy interpreter.
+
+    The tags consist of what is yielded by generic_tags() with a calcuated value
+    for 'interpreter' if it is not specified.
+    """
+    # The 'interpreter' parameter is necessary to allow the user to distinguish
+    # between PyPy and PyPy3 (pp and pp3, respectively).
+    if not interpreter:
+        interpreter = _pypy_interpreter()
+    for tag in generic_tags(interpreter, abis, platforms):
         yield tag
 
 
-def _generic_tags(interpreter, py_version, abi, platforms):
-    # type: (str, PythonVersion, str, Iterable[str]) -> Iterator[Tag]
-    for tag in (Tag(interpreter, abi, platform) for platform in platforms):
-        yield tag
-    if abi != "none":
-        tags = (Tag(interpreter, "none", platform_) for platform_ in platforms)
-        for tag in tags:
-            yield tag
+def _generic_interpreter(warn=False):
+    # type: (bool) -> str
+    version = _get_config_var("py_version_nodot", warn)
+    if not version:
+        version = "".join(map(str, sys.version_info[:2]))
+    return "{name}{version}".format(name=_interpreter_name(), version=version)
+
+
+def generic_tags(
+    interpreter=None,  # type: Optional[str]
+    abis=None,  # type: Optional[Iterable[str]]
+    platforms=None,  # type: Optional[Iterable[str]]
+    **kwargs  # type: bool
+):
+    # type: (...) -> Iterator[Tag]
+    """
+    Yields the tags for a generic interpreter.
+
+    The tags consist of:
+    - <interpreter>-<abi>-<platform>
+    - <intepreter>-none-<platform>  # If 'none' is not specified in 'abis'.
+
+    """
+    warn = _warn_parameter("generic_tags", kwargs)
+    if not interpreter:
+        interpreter = _generic_interpreter(warn=warn)
+    if not abis:
+        abis = [_generic_abi()]
+    platforms = list(platforms or _platforms())
+    abis = list(abis)
+    for abi in abis:
+        for platform_ in platforms:
+            yield Tag(interpreter, abi, platform_)
+    if "none" not in abis:
+        for platform_ in platforms:
+            yield Tag(interpreter, "none", platform_)
 
 
 def _py_interpreter_range(py_version):
     # type: (PythonVersion) -> Iterator[str]
     """
-    Yield Python versions in descending order.
+    Yields Python versions in descending order.
 
     After the latest version, the major-only version will be yielded, and then
-    all following versions up to 'end'.
+    all previous versions of that major version.
     """
     yield "py{major}{minor}".format(major=py_version[0], minor=py_version[1])
```

Comment:
This can also be done with `filter()`.

The key reason I'm hesitant is it will make `generic_tags()` different from `cpython_tags()` in terms of supporting a major-only `python_version` argument. And so documenting that unique distinction will have to be made very clear and some unique parameter name will be necessary so people don't assume there's a naming convention.

Reply:
So callers should provide a dummy value for minor version in both cases (`compatible_tags()` and `cpython_tags()`)?

Reply:
Fair point about the potential problem of future semantics. Unfortunately I still don't have a good answer for my other concerns either. 😞

Reply:
Is the problem that we can't come up with reasonable behavior in major and (major, minor) cases for both functions? I assume it would be whatever is generally applicable regardless of minor version.

Reply:
Correct. There is no concept of major-only CPython compatibility and so this doesn't really apply to `cpython_tags()` while it could make sense to `compatibility_tags()`. So that would make both having a parameter named `python_version` that actually have different type requirements a bit confusing. So we would need to come up with some other parameter name for one of them to make this work. Maybe `python_feature_version` for `cpython_tags()`?

Reply:
Wouldn't `cp3-abi3-*` make sense? Maybe that could be the return value for `cpython_tags()` if given a single-part version (which `== 3`), nothing otherwise, and we can avoid the need to decide on a reasonable-sounding parameter name. :)

Reply:
@chrahunt I'll think about it. I still think `cp3` is worthless, but if it will resolve this unfortunately design issue then I might be willing to just live with it under a pragmatic argument.